src/java.base/share/classes/java/lang/reflect/Proxy.java

```diff
@@ -1120,4 +1139,199 @@ public static InvocationHandler getInvocationHandler(Object proxy)
+            MethodHandles.Lookup lookup = MethodHandles.lookup();
+            Class<?> proxyInterface = findProxyInterfaceOrElseThrow(proxyClass, method);
+            MethodHandle dmh;
+            try {
+                dmh = proxyClassLookup(lookup, proxyClass)
+                        .findSpecial(proxyInterface, method.getName(), type, proxyClass)
+                        .withVarargs(false);
+            } catch (IllegalAccessException | NoSuchMethodException e) {
+                // should not reach here
+                throw new InternalError(e);
+            }
+            // this check can be turned into assertion as it is guaranteed to succeed by the virtue of
+            // looking up a default (instance) method declared or inherited by proxyInterface
+            // while proxyClass implements (is a subtype of) proxyInterface ...
+            assert ((BooleanSupplier) () -> {
+                try {
+                    // make sure that the method type matches
+                    dmh.asType(type.insertParameterTypes(0, proxyClass));
+                    return true;
+                } catch (WrongMethodTypeException e) {
+                    return false;
+                }
+            }).getAsBoolean() : "Wrong method type";
+            // change return type to Object
+            MethodHandle mh = dmh.asType(dmh.type().changeReturnType(Object.class));
+            // wrap any exception thrown with InvocationTargetException
+            mh = MethodHandles.catchException(mh, Throwable.class, InvocationException.wrapMH());
+            // spread array of arguments among parameters (skipping 1st parameter - target)
+            mh = mh.asSpreader(1, Object[].class, type.parameterCount());
+            // change target type to Object
+            mh = mh.asType(MethodType.methodType(Object.class, Object.class, Object[].class));
+
+            // push MH into cache
+            MethodHandle cached = methods.putIfAbsent(method, mh);
+            if (cached != null) {
+                superMH = cached;
+            } else {
+                superMH = mh;
+            }
+        }
+        return superMH;
+    }
+
+    /**
+     * Finds the first proxy interface that declares the given method
+     * directly or indirectly.
+     *
+     * @throws IllegalArgumentException if not found
+     */
+    private static Class<?> findProxyInterfaceOrElseThrow(Class<?> proxyClass, Method method) {
+        Class<?> declaringClass = method.getDeclaringClass();
+        if (!declaringClass.isInterface()) {
+            throw new IllegalArgumentException("\"" + method +
+                    "\" is not a method declared in the proxy class");
+        }
+
+        List<Class<?>> proxyInterfaces = Arrays.asList(proxyClass.getInterfaces());
+        // the method's declaring class is a proxy interface
+        if (proxyInterfaces.contains(declaringClass))
+            return declaringClass;
+
```

Comment:
It might be beneficial for future maintainers if you could include a comment here on the search order.

Reply:
What about:

```
+        // find the first proxy interface that inherits the default method
+        // i.e. the declaring class of the default method is a superinterface
+        // of the proxy interface
```